back/llm.py

```diff
@@ -0,0 +1,174 @@
+from abc import ABC, abstractmethod
+import asyncio
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from enum import Enum, auto
+from fastapi import FastAPI
+import openai
+import os
+from typing import Dict, Optional, Type
+
+app = FastAPI()
+
+class LLMProvider(str, Enum):
+    TEST = "test",
+    OPENAI = "openai"
+
+@dataclass
+class HeadlineRequest():
+    headline: str
+    author: str
+    body: str
+    provider: LLMProvider = LLMProvider.OPENAI  # Default to OpenAI if not specified
+    fallback_provider: Optional[LLMProvider] = LLMProvider.TEST  # Optional fallback provider
+
+@dataclass
+class HeadlineResponse():
+    original_headline: str
+    transformed_headline: str
+    provider_used: LLMProvider  # Tell the client which provider was actually used
+
+class LLM(ABC):
+    @abstractmethod
+    async def transform_headline(self, headline: str, author: str, body: str) -> str:
+        pass
+
+class TestLLM(LLM):
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        pass
```

Comment:
My understanding was that these magic methods are needed for Python async, but I think there's a way to just use the `@asynccontextmanager` decorator on the class to avoid having to define these manually, I'll see if I can get that working.

Reply:
Ah, I see. I thought they were just used to enable the `with` syntax. Will the code still work if you instantiate without the `with` and remove the magic methods?